### PR TITLE
Debug excel enable conversion of data and free text columns of input / output

### PR DESCRIPTION
**Pull request by Etschbeijer**
Minor fixes of smaller problems e.g. column sizes adapt correctly now, when a template is added or building block is deleted
Main problem could not be fully solved, requires fix in ARCtrl

### CODE REVIEW

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 1]
```diff
@@ -1762,7 +1804,7 @@ let deleteSelectedExcelColumns (selectedColumns: seq<int>) (excelTable: Table) =
 /// <param name="newBB"></param>
 /// <param name="table"></param>
 /// <param name="context"></param>
-let addBuildingBlockAt (excelIndex: int) (newBB: CompositeColumn) (table: Table) (context: RequestContext)=
+let addBuildingBlockCellAt (excelIndex: int) (newBB: CompositeColumn) (table: Table) (context: RequestContext) =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1807]: Wyh the rename? This adds a CompositeColumn right? therefore it adds a building block and not a cell

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 2]
```diff
@@ -1791,6 +1833,43 @@ let addBuildingBlockAt (excelIndex: int) (newBB: CompositeColumn) (table: Table)
         |> List.iteri (fun ci header -> ExcelHelper.addColumnAndRows (float (excelIndex + ci)) table header bodyValues.[ci] |> ignore)
     }
 
+type BodyCellType =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1836]: This type will be added to Shared as CompositeCellDiscriminate in my pr #549

Comment (Freymaurer) [reply src/Client/OfficeInterop/OfficeInterop.fs:1836]: keep it here and add a [<System.Obsolete("Replace with Shared CompositeCellDiscriminate")>]

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 3]
```diff
@@ -1791,6 +1833,43 @@ let addBuildingBlockAt (excelIndex: int) (newBB: CompositeColumn) (table: Table)
...
+            | Result.Ok excelTable ->
+                let! _, rowIndex = getSelectedBuildingBlockCell excelTable context
+                let! arcTable = ArcTable.tryGetFromExcelTable(excelTable, context)
+                let! (arcMainColumn, _) = getArcMainColumn excelTable arcTable.Value context
+
+                if (arcMainColumn.Validate false) && rowIndex > 0 then
+
+                    let result =
+                        match arcMainColumn with
+                        | amc when amc.Header.ToString() = "Output [Sample Name]" -> None
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1857]: I might not understand this part, but it seems to be very bad design to match to specific strings if we want to rely on ARCtrl for newest ARC specification changes! 

In addition, for example how is `Output [Material]` handled?

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 4]
```diff
@@ -1791,6 +1833,43 @@ let addBuildingBlockAt (excelIndex: int) (newBB: CompositeColumn) (table: Table)
...
+        promise {
+            let! result = tryGetActiveAnnotationTable context
+
+            match result with
+            | Result.Ok excelTable ->
+                let! _, rowIndex = getSelectedBuildingBlockCell excelTable context
+                let! arcTable = ArcTable.tryGetFromExcelTable(excelTable, context)
+                let! (arcMainColumn, _) = getArcMainColumn excelTable arcTable.Value context
+
+                if (arcMainColumn.Validate false) && rowIndex > 0 then
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1853]: You read the table in line 1850, the table and therefore all columns must be valid.

**Review by Freymaurer — CHANGES_REQUESTED** [review]

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 1 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs:1807]: I doubled he old method for testing and forgot to rename it afterwards

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 2 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 1 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 3 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs (hunk 4 above)
Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 5]
```diff
@@ -1850,17 +1837,15 @@ let getSelectedCellType () =
...
-                        | amc when amc.Header.ToString() = "Input [Source Name]" -> None
-                        | amc when amc.Cells.[(int rowIndex) - 1].isUnitized -> Some BodyCellType.Unitized
-                        | amc when amc.Cells.[(int rowIndex) - 1].isTerm -> Some BodyCellType.Term
-                        | amc when amc.Cells.[(int rowIndex) - 1].isData -> Some BodyCellType.Data
-                        | amc when amc.Header.isInput && amc.Cells.[(int rowIndex) - 1].isFreeText -> Some BodyCellType.Text
-                        | amc when amc.Header.isOutput && amc.Cells.[(int rowIndex) - 1].isFreeText -> Some BodyCellType.Text
+                        | amc when amc.Cells.[(int rowIndex) - 1].isUnitized -> Some CompositeCellDiscriminate.Unitized
+                        | amc when amc.Cells.[(int rowIndex) - 1].isTerm -> Some CompositeCellDiscriminate.Term
+                        | amc when amc.Cells.[(int rowIndex) - 1].isData -> Some CompositeCellDiscriminate.Data
+                        | amc when amc.Header.isInput && amc.Header.IsDataColumn && amc.Cells.[(int rowIndex) - 1].isFreeText -> Some CompositeCellDiscriminate.Text
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1847]: why are you checking headers for this again? Can you not simply check if cell _.isFreetext?

**Review by Freymaurer — CHANGES_REQUESTED** [review]

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 6]
```diff
@@ -1633,6 +1631,68 @@ let getSelectedBuildingBlockCell (table: Table) (context: RequestContext) =
...
+        promise {
+            let! result = tryGetActiveAnnotationTable context
+            match result with
+            | Result.Ok excelTable ->
+                let! arcTable = ArcTable.tryGetFromExcelTable(excelTable, context)
+                let! (arcMainColumn, _) = getArcMainColumn excelTable arcTable.Value context
+
+                return
+                    match cellType with
+                    | Some CompositeCellDiscriminate.Unitized -> Some CompositeCellDiscriminate.Term
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:1683]: This function must be reworked.

Here is what you do as far as i understand it:

goal: Return valid celltypes for conversion based on `cellType`

1. get active table
2. try parse to arctable, then take `.Value` without error handling
3. get selected main column
4. check if valid conversion

BUT;
- if `cellType` is term return unitized without any checking against the selected building block
- Similiar for `cellType` = Data

Comment (Etschbeijer) [reply src/Client/OfficeInterop/OfficeInterop.fs]: fixed (hopefully)

**Review by Freymaurer — APPROVED** [review]

**Pull request merged**